horizon/facts/update_subscriber.py

```diff
@@ -0,0 +1,95 @@
+import asyncio
+
+from loguru import logger
+from opal_client.data.updater import DataUpdater
+from opal_common.schemas.data import DataSourceEntry
+
+
+class DataUpdateSubscriber:
+    def __init__(self, updater: DataUpdater):
+        self._updater = updater
+        self._topic_events: dict[str, asyncio.Event] = {}
+
+    @property
+    def callbacks(self) -> dict[str, list[callable]]:
+        # Callbacks registered for PubSub client by topics
+        return self._updater._client._callbacks  # noqa
+
+    async def _on_message(self, topic: str = "", data=None) -> None:
+        if topic in self._topic_events:
+            logger.debug(f"Resolving subscriber event for topic: {topic}")
+            self._topic_events[topic].set()
```

Comment:
you can have multiple events on the same topic and this would mark all of them as "finished", am I wrong ?

Reply:
Right, we wait for the next update arriving for this topic, regardless if its caused by our publish or not

Reply:
Would it really achieve the goal of waiting for a specific update ? 
I'm imagining a PDP that receives a lot of update, this wouldn't really solve this for it.
IMO the event waiting should be by topic & update_id

Reply:
OPAL does not have this capability, and I cannot think of a way of extending it easily without changing OPAL internally

Reply:
it is possible to wait for a specific update, and that was the original intention.
when you publish a `DataUpdate` message, it can receive an optional `id`, and when you get this id back in the client - at this point you are triggering the event.

see here:
https://github.com/permitio/opal/blob/fc23168de2573671063e3cb280284681f63ed04e/packages/opal-common/opal_common/schemas/data.py#L151

Reply:
according to [this](https://github.com/permitio/fastapi_websocket_pubsub/blob/master/fastapi_websocket_pubsub/pub_sub_client.py#L318) you can add a callback to a "magic" topic that will add it to all topics ( ALL_TOPICS ).
and I think the callback includes also the data provided ( which might include the update id but not sure about it )